src/address.rs

```diff
@@ -0,0 +1,155 @@
+// Address conversion utilities based on the Pontem address solution.
+// To properly handle Move VM addresses and Substrate addresses, we need to convert them to each other.
+
+use codec::{Decode, Encode, Error};
+use move_core_types::account_address::AccountAddress;
+
+/// Convert Move VM address instance (AccountAddress) to an AccountId.
+///
+/// Returns an AccountId instance or decoding error.
+pub fn address_to_account<AccountId>(address: &AccountAddress) -> Result<AccountId, Error>
+where
+    AccountId: Decode + Sized,
+{
+    AccountId::decode(&mut address.as_ref())
+}
+
+/// Convert AccountId to Move VM address format.
+///
+/// Returns an AccountAddress instance - shouldn't fail as any bytes with length 32 could be
+/// represented as an AccountAddress.
+///
+/// ```
+/// use pallet_move::address::account_to_address;
+/// use sp_core::sr25519::Public;
+/// use sp_core::crypto::Ss58Codec;
+///
+/// let address = account_to_address(& Public::from_ss58check_with_version("gkQ5K6EnLRgZkwozG8GiBAEnJyM6FxzbSaSmVhKJ2w8FcK7ih").unwrap().0);
+/// assert_eq!(address.to_string(), "d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d");
+/// ```
+pub fn account_to_address<AccountId>(account: &AccountId) -> AccountAddress
+where
+    AccountId: Encode,
+{
+    AccountAddress::new(account_to_bytes(account))
+}
+
+/// Convert AccountId to byte format which is compatible with Move VM address.
+///
+/// In fact, returned value can be just passed to the AccountAddress constructor and it will
+/// create a valid Move VM address.
+/// This function is ready to handle different than 32 bytes long AccountId instances.
+/// For the MoveVM port, we need to have a 32 bytes long address.
+///
+/// ```
+/// use pallet_move::address::account_to_bytes;
+/// use sp_core::sr25519::Public;
+/// use sp_core::crypto::Ss58Codec;
+///
+/// let bytes = account_to_bytes(& Public::from_ss58check_with_version("gkQ5K6EnLRgZkwozG8GiBAEnJyM6FxzbSaSmVhKJ2w8FcK7ih").unwrap().0);
+/// assert_eq!(bytes.as_ref(), &[0xD4, 0x35, 0x93, 0xC7, 0x15, 0xFD, 0xD3, 0x1C, 0x61, 0x14, 0x1A, 0xBD, 0x04, 0xA9, 0x9F, 0xD6,
+///                 0x82, 0x2C, 0x85, 0x58, 0x85, 0x4C, 0xCD, 0xE3, 0x9A, 0x56, 0x84, 0xE7, 0xA5, 0x6D, 0xA2, 0x7D]);
+/// ```
+pub fn account_to_bytes<AccountId>(account: &AccountId) -> [u8; AccountAddress::LENGTH]
+where
+    AccountId: Encode,
+{
+    const LENGTH: usize = AccountAddress::LENGTH;
+    let mut result = [0; LENGTH];
+    let bytes = account.encode();
+
+    let skip = LENGTH.saturating_sub(bytes.len());
+
+    result[skip..].copy_from_slice(&bytes);
```

Comment:
We could panic here if `bytes.len()` is larger than 32 bytes (since slices won't have equal sizes).

And since the generic input argument `account` only needs to implement the `Encode` trait, there is no guarantee it won't be larger than 32 bytes once encoded.

Maybe we can handle such cases - we don't want anybody to use these functions with something that's not an actual account. And then we could write a test for that here.

Reply:
For now, I've no better idea to narrow it using any trait, but I wouldn't return `Result` instead of actual bytes - it would introduce additional overhead for getting a byte array.

I would also avoid cutting `bytes` value to fit into the result length as it may introduce some unwanted results.

Doc has been updated to include information about it, and a test has been added.

Reply:
> but I wouldn't return Result

A documented panic is then okay to use here.

Maybe a comment above the `copy_from_slice` line would be good to have.

Or optimally, check the lengths before doing `copy_from_slice`  and then panic with a more informative description.

Reply:
Added assertion.